docs/reference.md

```diff
@@ -131,3 +131,66 @@ dedicated test-suite, consisting of several tests exercising various runtime
 behaviors, for each StableHLO Op. The tests can be found
 [here](https://github.com/openxla/stablehlo/tree/main/stablehlo/tests/) (e.g.
 interpret\_\*.mlir).
+
+### Testing guidelines
+
+**(G1) Do we need to test for all the supported types for every op?**
+
+We can use a combination of following rules to decide it:
+
+  1. While implementing an op, if there exists code in the corresponding `eval`
+     function to handle a particular type, then it is imperative to have test(s)
+     to cover for that type. As an example, for `add` op, there is exclusive
+     code to handle Integer, Boolean, Floating-point, and Complex types, and hence we
+     need one test for each type.
+  2. If a set of types are handled uniformly in the corresponding
+     `eval` function, then a single test for all those types should be
+     sufficient. As an example, for `add` op, all the variants of Integers
+     types (`si4`, `u4`, `si8`, `u8` and so on) are handled alike using
+     `llvm::APInt` APIs, and hence we can skip adding tests for each of those
+     variants, and instead, add a single representative test. To avoid ambiguity
+     in selecting the representative, we should use the following guidelines:
+
+       * If all the types, handled uniformaly, have the same primitive type
+         (i.e., if all are Integer, or Floating-point, or Complex types), then
+         choose the one with maximum bit-width.
+       * If all the types, handled uniformaly, have a mix of primitive types,
+         then choose the one with the following primitive type, in decreasing
+         order of preference: Boolean, Integer, Floating-point, Complex.
+
+**(G2) How about adding tests dedicated for testing the interpreter infrastructure?**
+
+The interpreter infrastructure is mostly straightforward and can be added it in
+our trust base. The only non-trivial part is how various types are packed into
+and unpacked from the underlying interpreter storage. As discussed in (G1), we
+will be testing only those types of an op which are handled differently. With
+that is possible that the packing/un-packing code, corresponding to different
+variants of Integer/Floating-point types, might not get fully covered during
+testing.
+
+To ensure that we can chose an op, like `add`, which supports all the StableHLO
+element types and write exhaustive tests. Fortunately, we already have it.
+
+**(G3) If the implementation of an op depends other ops, should be write tests for the latter?**
+
+No. For example, the implementation of `batch_norm_grad` can be based on
+`divide`, `subtract`, `multiply` and others, we should avoid testing the latter
+ops while testing the former.
+
+**(G4) Should we write tests to exercise the implementation-defined / undefined behaviors?**
+
+We should not write tests which exercises the implementation defined or
```

Comment:
Typo: "exercise". (As well as "demonstrate" two lines below and "do not" three lines below).